File: advent_of_code/2025/day_01.go
package main

import (
	"fmt"
	"io"
	"net/http"
	"os"
	"strconv"
	"strings"
)

const baseURL = "https://adventofcode.com/2025/day/1/input"

type Solution struct {
	HTTPClient *http.Client
	Session    string
	Input      string
	Password   int
}

func NewSolution(session string) *Solution {
	return &Solution{
		HTTPClient: http.DefaultClient,
		Session:    session,
	}
}

func (s *Solution) GetInput() error {
	req, err := http.NewRequest(http.MethodGet, baseURL, nil)
	if err != nil {
		return err
	}
	req.AddCookie(&http.Cookie{Name: "session", Value: s.Session})
	resp, err := s.HTTPClient.Do(req)
	if err != nil {
		return err
	}
	defer resp.Body.Close()
	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return err
	}
	s.Input = string(body)
	return nil
}

func mod100(x int) int {
	m := x % 100
	if m < 0 {
		m += 100
	}
	return m
}

func (s *Solution) GetFirstSolution() {
	current := 50
	total := 0
	for line := range strings.SplitSeq(strings.TrimSpace(s.Input), "\n") {
		if line == "" {
			continue
		}
		goLeft := line[0] == 'L'
		steps, _ := strconv.Atoi(line[1:])
		if goLeft {
			current = mod100(current - steps)
		} else {
			current = mod100(current + steps)
		}
		if current == 0 {
			total++
		}
	}
	s.Password = total
}

func (s *Solution) GetSecondSolution() {
	current := 50
	total := 0
	for line := range strings.SplitSeq(strings.TrimSpace(s.Input), "\n") {
		if line == "" {
			continue
		}
		goLeft := line[0] == 'L'
		steps, _ := strconv.Atoi(line[1:])
		for k := 1; k <= steps; k++ {
			var pos int
			if goLeft {
				pos = mod100(current - k)
			} else {
				pos = mod100(current + k)
			}
			if pos == 0 {
				total++
			}
		}
		if goLeft {
			current = mod100(current - steps)
		} else {
			current = mod100(current + steps)
		}
	}
	s.Password = total
}

func main() {
	session := os.Getenv("AOC_SESSION")
	if session == "" {
		fmt.Fprintln(os.Stderr, "set AOC_SESSION to your adventofcode.com session cookie value")
		os.Exit(1)
	}
	solution := NewSolution(session)
	if err := solution.GetInput(); err != nil {
		fmt.Fprintln(os.Stderr, err)
		os.Exit(1)
	}
	// solution.GetFirstSolution()
	// fmt.Println(solution.Password)
	solution.GetSecondSolution()
	fmt.Println(solution.Password)
}

File: advent_of_code/2025/day_01.py
import browser_cookie3
import requests

class Solution:
    BASE_URL = 'https://adventofcode.com/2025/day/1/input'

    def __init__(self, session_cookie):
        self._session_cookie = session_cookie

    def get_input(self):
        response = requests.get(self.BASE_URL, cookies=self._session_cookie)
        self._input = response.text

    def get_first_solution(self):
        current_position = 50
        total_zeroes = 0
        for instruction in self._input.strip().split("\n"):
            go_left = instruction[0] == "L"
            steps = int(instruction[1:])

            if go_left:
                current_position = (current_position - steps) % 100
            else:
                current_position = (current_position + steps) % 100
            
            if current_position == 0:
                total_zeroes += 1
        
        self.password = total_zeroes

    def get_second_solution(self):
        current_position = 50
        total_zeroes = 0
        for instruction in self._input.strip().split("\n"):
            go_left = instruction[0] == "L"
            steps = int(instruction[1:])

            for k in range(1, steps + 1):
                if go_left:
                    pos = (current_position - k) % 100
                else:
                    pos = (current_position + k) % 100
                if pos == 0:
                    total_zeroes += 1

            if go_left:
                current_position = (current_position - steps) % 100
            else:
                current_position = (current_position + steps) % 100

        self.password = total_zeroes

if __name__ == "__main__":
    # Grab cookies for the domain
    cj = browser_cookie3.chrome(domain_name="adventofcode.com")

    solution = Solution(cj)
    solution.get_input()
    # solution.get_first_solution()
    # print(solution.password)
    solution.get_second_solution()
    print(solution.password)
